File: examples/word-page-numbers.ts
/**
 * 例子（Epic-34）：演示在正文里插入 PAGE / NUMPAGES 字段。
 *
 * 跑法：
 *   bun run examples/word-page-numbers.ts <output.docx>
 *
 * 注：完整页脚 wiring 需要 FooterPart + sectPr 关联，本例仅在正文段落里展示
 * 字段；Word 打开时会自动算出当前页码与总页数。
 */

import {
  Paragraph,
  Run,
  Text,
  WordprocessingDocument,
  createPageNumberRun,
  createTotalPagesRun,
} from "../src/word/index.js";

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: word-page-numbers <output.docx>\n");
    process.exit(2);
  }

  const doc = WordprocessingDocument.create();
  const body = doc.mainDocumentPart?.document.firstChild()!;
  const append = (e: unknown): void =>
    (body as { appendChild: (e: unknown) => void }).appendChild(e);

  function addPlain(text: string): void {
    const p = new Paragraph();
    const r = new Run();
    const t = new Text();
    t.text = text;
    r.appendChild(t);
    p.appendChild(r);
    append(p);
  }

  addPlain("This document demonstrates PAGE and NUMPAGES fields.");
  addPlain("Open in Word and the next paragraph should show the live values.");

  const p = new Paragraph();
  const lead = new Run();
  const lt = new Text();
  lt.text = "第 ";
  lead.appendChild(lt);
  p.appendChild(lead);

  p.appendChild(createPageNumberRun());

  const mid = new Run();
  const mt = new Text();
  mt.text = " 页 / 共 ";
  mid.appendChild(mt);
  p.appendChild(mid);

  p.appendChild(createTotalPagesRun());

  const trail = new Run();
  const tt = new Text();
  tt.text = " 页";
  trail.appendChild(tt);
  p.appendChild(trail);

  append(p);

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Wrote ${outputPath} (PAGE + NUMPAGES fields)\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/WordPageNumbers.cs
// Replica of examples/word-page-numbers.ts
// Creates a docx with 2 plain paragraphs + a paragraph with PAGE and NUMPAGES fields.
// Uses the fldChar sequence (begin / instrText / separate / end) matching the TS
// createPageNumberRun / createTotalPagesRun helpers.

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;

namespace CrossSdkVerify.Replicas;

public static class WordPageNumbers
{
    public static void Run(string outputPath)
    {
        using var doc = WordprocessingDocument.Create(outputPath,
            DocumentFormat.OpenXml.WordprocessingDocumentType.Document);

        var mainPart = doc.AddMainDocumentPart();
        mainPart.Document = new Document();
        var body = new Body();
        mainPart.Document.Append(body);

        body.Append(new Paragraph(new Run(new Text("This document demonstrates PAGE and NUMPAGES fields."))));
        body.Append(new Paragraph(new Run(new Text("Open in Word and the next paragraph should show the live values."))));

        // Paragraph: "第 " + PAGE field + " 页 / 共 " + NUMPAGES field + " 页"
        var p = new Paragraph();

        p.Append(new Run(new Text("第 ") { Space = SpaceProcessingModeValues.Preserve }));

        // PAGE field
        AppendField(p, "PAGE");

        p.Append(new Run(new Text(" 页 / 共 ") { Space = SpaceProcessingModeValues.Preserve }));

        // NUMPAGES field
        AppendField(p, "NUMPAGES");

        p.Append(new Run(new Text(" 页") { Space = SpaceProcessingModeValues.Preserve }));

        body.Append(p);
        body.Append(new SectionProperties());
        mainPart.Document.Save();
    }

    private static void AppendField(Paragraph p, string fieldName)
    {
        // Run 1: fldChar begin
        p.Append(new Run(new FieldChar { FieldCharType = FieldCharValues.Begin }));
        // Run 2: instrText
        p.Append(new Run(new FieldCode($" {fieldName} ") { Space = SpaceProcessingModeValues.Preserve }));
        // Run 3: fldChar separate
        p.Append(new Run(new FieldChar { FieldCharType = FieldCharValues.Separate }));
        // Run 4: fldChar end
        p.Append(new Run(new FieldChar { FieldCharType = FieldCharValues.End }));
    }
}
